[ICE]: `expected a projection`

<!--
[31mICE[0m: Rustc ./DC87D749F7189DB6749EB7B0F246583DCA01BDEC0317D86213F29E3CDF421FE8.rs '-Znext-solver=globally -ooutputfile -Zdump-mir-dir=dir' 'thread 'rustc' ($TID) panicked at compiler/rustc_type_ir/src/predicate.rs:689:9: 'expected a projection'', 'thread 'rustc' ($TID) panicked at compiler/rustc_type_ir/src/predicate.rs:689:9: 'expected a projection''
File: /tmp/im/DC87D749F7189DB6749EB7B0F246583DCA01BDEC0317D86213F29E3CDF421FE8.rs
-->

auto-reduced (treereduce-rust):
````rust
//@compile-flags: -Znext-solver=globally
type Foo = Vec<impl Send>;

#[define_opaque(Foo)]
fn make_foo() -> Foo {}

````

original:
````rust
#![feature(type_alias_impl_trait)]

#[define_opaque(Foo)]
fn make_foo(Vec: Vec<dyn Fn() -> Vec<dyn Fn() -> Vec<dyn Fn() -> Vec<dyn Fn() -> Vec<dyn Fn() -> ()>>>>>) -> Foo {
    vec![true, false]
}
#![feature(type_alias_impl_trait)]

type Foo = Vec<impl Send>;

#[define_opaque(Foo)]
fn make_foo() -> Foo {}

fn main() {}

````

Version information
````
rustc 1.95.0-nightly (621d76794 2026-01-18)
binary: rustc
commit-hash: 621d76794c76fc21c0a6151fbc090120e0230a91
commit-date: 2026-01-18
host: x86_64-unknown-linux-gnu
release: 1.95.0-nightly
LLVM version: 21.1.8

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/621d76794c76fc21c0a6151fbc090120e0230a91/compiler/rustc_type_ir/src/predicate.rs#L683-L695

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc -Znext-solver=globally`

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0658]: use of unstable library feature `type_alias_impl_trait`: `type_alias_impl_trait` has open design concerns
 --> /tmp/icemaker_global_tempdir.ZPVHCxSfZFnB/rustc_testrunner_tmpdir_reporting.gzsYmEyjuNkQ/mvce.rs:3:3
  |
3 | #[define_opaque(Foo)]
  |   ^^^^^^^^^^^^^
  |
  = note: see issue #63063 <https://github.com/rust-lang/rust/issues/63063> for more information
  = help: add `#![feature(type_alias_impl_trait)]` to the crate attributes to enable
  = note: this compiler was built on 2026-01-18; consider upgrading it if it is out of date

error[E0658]: `impl Trait` in type aliases is unstable
 --> /tmp/icemaker_global_tempdir.ZPVHCxSfZFnB/rustc_testrunner_tmpdir_reporting.gzsYmEyjuNkQ/mvce.rs:1:16
  |
1 | type Foo = Vec<impl Send>;
  |                ^^^^^^^^^
  |
  = note: see issue #63063 <https://github.com/rust-lang/rust/issues/63063> for more information
  = help: add `#![feature(type_alias_impl_trait)]` to the crate attributes to enable
  = note: this compiler was built on 2026-01-18; consider upgrading it if it is out of date

error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.ZPVHCxSfZFnB/rustc_testrunner_tmpdir_reporting.gzsYmEyjuNkQ/mvce.rs:4:24
  |
4 | fn make_foo() -> Foo {}
  |                        ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.ZPVHCxSfZFnB/rustc_testrunner_tmpdir_reporting.gzsYmEyjuNkQ/mvce.rs`


thread 'rustc' (2113122) panicked at /rustc-dev/621d76794c76fc21c0a6151fbc090120e0230a91/compiler/rustc_type_ir/src/predicate.rs:689:9:
expected a projection
stack backtrace:
   0:     0x7f460c233fd3 - <<std[287527946f71f0bc]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[7477a31a1a34cc49]::fmt::Display>::fmt
   1:     0x7f460c80e2c8 - core[7477a31a1a34cc49]::fmt::write
   2:     0x7f460c24a9c6 - <std[287527946f71f0bc]::sys::stdio::unix::Stderr as std[287527946f71f0bc]::io::Write>::write_fmt
   3:     0x7f460c20a048 - std[287527946f71f0bc]::panicking::default_hook::{closure#0}
   4:     0x7f460c227703 - std[287527946f71f0bc]::panicking::default_hook
   5:     0x7f460b1fbb8a - std[287527946f71f0bc]::panicking::update_hook::<alloc[979041e9952e771b]::boxed::Box<rustc_driver_impl[517260a07841a630]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7f460c2279e2 - std[287527946f71f0bc]::panicking::panic_with_hook
   7:     0x7f460c20a13a - std[287527946f71f0bc]::panicking::panic_handler::{closure#0}
   8:     0x7f460c200dc9 - std[287527946f71f0bc]::sys::backtrace::__rust_end_short_backtrace::<std[287527946f71f0bc]::panicking::panic_handler::{closure#0}, !>
   9:     0x7f460c20bbdd - __rustc[5ec7081ce87b032f]::rust_begin_unwind
  10:     0x7f4608d0cfbc - core[7477a31a1a34cc49]::panicking::panic_fmt
  11:     0x7f460aa8aed1 - <rustc_type_ir[b93cf349b3a4f34d]::predicate::AliasTerm<rustc_middle[725063e074f29df]::ty::context::TyCtxt>>::trait_def_id
  12:     0x7f460ab26206 - <rustc_trait_selection[9fee51d8a10a777c]::error_reporting::TypeErrCtxt>::report_fulfillment_errors
  13:     0x7f460cc0e926 - <rustc_hir_typeck[ed5453188701a25b]::fn_ctxt::FnCtxt>::try_structurally_resolve_type
  14:     0x7f460c8d6eab - <rustc_hir_typeck[ed5453188701a25b]::fn_ctxt::FnCtxt>::coerce
  15:     0x7f460cbfa346 - rustc_hir_typeck[ed5453188701a25b]::check::check_fn
  16:     0x7f460d5b3d33 - rustc_hir_typeck[ed5453188701a25b]::typeck_with_inspect::{closure#0}
  17:     0x7f460d5b26b4 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>>
  18:     0x7f460c95522a - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_data_structures[e0d8483206daafa2]::vec_cache::VecCache<rustc_span[851f35e07e3c2dd4]::def_id::LocalDefId, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[b7e73caa2c2ad9a]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  19:     0x7f460c954b09 - rustc_query_impl[f20b8284200e680a]::query_impl::typeck::get_query_non_incr::__rust_end_short_backtrace
  20:     0x7f460ce5010a - rustc_mir_build[80b6bf0f2af09b99]::thir::pattern::check_match::check_match
  21:     0x7f460ce4f4d1 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::check_match::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 1usize]>>
  22:     0x7f460cf92f78 - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_data_structures[e0d8483206daafa2]::vec_cache::VecCache<rustc_span[851f35e07e3c2dd4]::def_id::LocalDefId, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 1usize]>, rustc_query_system[b7e73caa2c2ad9a]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  23:     0x7f460cf92c3d - rustc_query_impl[f20b8284200e680a]::query_impl::check_match::get_query_non_incr::__rust_end_short_backtrace
  24:     0x7f460cd11667 - rustc_mir_build[80b6bf0f2af09b99]::builder::build_mir_inner_impl
  25:     0x7f460c808dea - rustc_mir_transform[f63ee9052e691c]::mir_built
  26:     0x7f460c808db3 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::mir_built::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>>
  27:     0x7f460c95522a - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_data_structures[e0d8483206daafa2]::vec_cache::VecCache<rustc_span[851f35e07e3c2dd4]::def_id::LocalDefId, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[b7e73caa2c2ad9a]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  28:     0x7f460c954c89 - rustc_query_impl[f20b8284200e680a]::query_impl::mir_built::get_query_non_incr::__rust_end_short_backtrace
  29:     0x7f460c84eb97 - rustc_mir_transform[f63ee9052e691c]::ffi_unwind_calls::has_ffi_unwind_calls
  30:     0x7f460c84e5a9 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::has_ffi_unwind_calls::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 1usize]>>
  31:     0x7f460cf92f78 - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_data_structures[e0d8483206daafa2]::vec_cache::VecCache<rustc_span[851f35e07e3c2dd4]::def_id::LocalDefId, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 1usize]>, rustc_query_system[b7e73caa2c2ad9a]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  32:     0x7f460cf92afd - rustc_query_impl[f20b8284200e680a]::query_impl::has_ffi_unwind_calls::get_query_non_incr::__rust_end_short_backtrace
  33:     0x7f460d526bcd - rustc_mir_transform[f63ee9052e691c]::mir_promoted
  34:     0x7f460d5269e0 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::mir_promoted::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 16usize]>>
  35:     0x7f460d1e9475 - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_data_structures[e0d8483206daafa2]::vec_cache::VecCache<rustc_span[851f35e07e3c2dd4]::def_id::LocalDefId, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 16usize]>, rustc_query_system[b7e73caa2c2ad9a]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  36:     0x7f460d1e900c - rustc_query_impl[f20b8284200e680a]::query_impl::mir_promoted::get_query_non_incr::__rust_end_short_backtrace
  37:     0x7f460d93d325 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::mir_borrowck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>>
  38:     0x7f460c95522a - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_data_structures[e0d8483206daafa2]::vec_cache::VecCache<rustc_span[851f35e07e3c2dd4]::def_id::LocalDefId, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[b7e73caa2c2ad9a]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  39:     0x7f460c94bf99 - rustc_query_impl[f20b8284200e680a]::query_impl::mir_borrowck::get_query_non_incr::__rust_end_short_backtrace
  40:     0x7f460b3a2023 - <rustc_hir_analysis[1a80394817eddfdd]::collect::type_of::opaque::TaitConstraintLocator>::check
  41:     0x7f460b3e2a50 - rustc_hir_analysis[1a80394817eddfdd]::collect::type_of::opaque::find_opaque_ty_constraints_for_tait
  42:     0x7f460e745274 - rustc_hir_analysis[1a80394817eddfdd]::collect::type_of::type_of_opaque.cold
  43:     0x7f460d813f47 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::type_of_opaque::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>>
  44:     0x7f460c8173c5 - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_query_system[b7e73caa2c2ad9a]::query::caches::DefIdCache<rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  45:     0x7f460d1d90db - rustc_query_impl[f20b8284200e680a]::query_impl::type_of_opaque::get_query_non_incr::__rust_end_short_backtrace
  46:     0x7f460d1bd8b6 - rustc_hir_analysis[1a80394817eddfdd]::collect::type_of::type_of
  47:     0x7f460c975c26 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::type_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>>
  48:     0x7f460c8173c5 - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_query_system[b7e73caa2c2ad9a]::query::caches::DefIdCache<rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  49:     0x7f460c816f6f - rustc_query_impl[f20b8284200e680a]::query_impl::type_of::get_query_non_incr::__rust_end_short_backtrace
  50:     0x7f460d814abb - rustc_hir_analysis[1a80394817eddfdd]::check::check::check_opaque
  51:     0x7f460cf9b904 - rustc_hir_analysis[1a80394817eddfdd]::check::check::check_item_type
  52:     0x7f460cf9379c - rustc_hir_analysis[1a80394817eddfdd]::check::wfcheck::check_well_formed
  53:     0x7f460cf9376b - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::check_well_formed::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 1usize]>>
  54:     0x7f460cf931e3 - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_data_structures[e0d8483206daafa2]::vec_cache::VecCache<rustc_span[851f35e07e3c2dd4]::def_id::LocalDefId, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 1usize]>, rustc_query_system[b7e73caa2c2ad9a]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  55:     0x7f460cf92cd6 - rustc_query_impl[f20b8284200e680a]::query_impl::check_well_formed::get_query_non_incr::__rust_end_short_backtrace
  56:     0x7f460cf904d1 - rustc_hir_analysis[1a80394817eddfdd]::check::wfcheck::check_type_wf
  57:     0x7f460cf9035d - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::check_type_wf::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 1usize]>>
  58:     0x7f460d9a3e85 - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_query_system[b7e73caa2c2ad9a]::query::caches::SingleCache<rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  59:     0x7f460d9a3b36 - rustc_query_impl[f20b8284200e680a]::query_impl::check_type_wf::get_query_non_incr::__rust_end_short_backtrace
  60:     0x7f460d33f316 - rustc_hir_analysis[1a80394817eddfdd]::check_crate
  61:     0x7f460c9493db - rustc_interface[5c48368f5debe80d]::passes::analysis
  62:     0x7f460c949035 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 0usize]>>
  63:     0x7f460d9a4be3 - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_query_system[b7e73caa2c2ad9a]::query::caches::SingleCache<rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  64:     0x7f460d9a49ce - rustc_query_impl[f20b8284200e680a]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  65:     0x7f460db82c11 - <rustc_interface[5c48368f5debe80d]::passes::create_and_enter_global_ctxt<core[7477a31a1a34cc49]::option::Option<rustc_interface[5c48368f5debe80d]::queries::Linker>, rustc_driver_impl[517260a07841a630]::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core[7477a31a1a34cc49]::ops::function::FnOnce<(&rustc_session[eb79e6d3104f29ae]::session::Session, rustc_middle[725063e074f29df]::ty::context::CurrentGcx, alloc[979041e9952e771b]::sync::Arc<rustc_data_structures[e0d8483206daafa2]::jobserver::Proxy>, &std[287527946f71f0bc]::sync::once_lock::OnceLock<rustc_middle[725063e074f29df]::ty::context::GlobalCtxt>, &rustc_data_structures[e0d8483206daafa2]::sync::worker_local::WorkerLocal<rustc_middle[725063e074f29df]::arena::Arena>, &rustc_data_structures[e0d8483206daafa2]::sync::worker_local::WorkerLocal<rustc_hir[7eac7792dedc1861]::Arena>, rustc_driver_impl[517260a07841a630]::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  66:     0x7f460d9f5d34 - rustc_interface[5c48368f5debe80d]::interface::run_compiler::<(), rustc_driver_impl[517260a07841a630]::run_compiler::{closure#0}>::{closure#1}
  67:     0x7f460d95aa3a - std[287527946f71f0bc]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[5c48368f5debe80d]::util::run_in_thread_with_globals<rustc_interface[5c48368f5debe80d]::util::run_in_thread_pool_with_globals<rustc_interface[5c48368f5debe80d]::interface::run_compiler<(), rustc_driver_impl[517260a07841a630]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  68:     0x7f460d95a7fe - <std[287527946f71f0bc]::thread::lifecycle::spawn_unchecked<rustc_interface[5c48368f5debe80d]::util::run_in_thread_with_globals<rustc_interface[5c48368f5debe80d]::util::run_in_thread_pool_with_globals<rustc_interface[5c48368f5debe80d]::interface::run_compiler<(), rustc_driver_impl[517260a07841a630]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[7477a31a1a34cc49]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  69:     0x7f460d95c5c4 - <std[287527946f71f0bc]::sys::thread::unix::Thread>::new::thread_start
  70:     0x7f460729698b - <unknown>
  71:     0x7f460731a9cc - <unknown>
  72:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.95.0-nightly (621d76794 2026-01-18) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z next-solver=globally -Z dump-mir-dir=dir

query stack during panic:
#0 [typeck] type-checking `make_foo`
#1 [check_match] match-checking `make_foo`
#2 [mir_built] building MIR for `make_foo`
#3 [has_ffi_unwind_calls] checking if `make_foo` contains FFI-unwind calls
#4 [mir_promoted] promoting constants in MIR for `make_foo`
#5 [mir_borrowck] borrow-checking `make_foo`
#6 [type_of_opaque] computing type of opaque `Foo::{opaque#0}`
#7 [type_of] computing type of `Foo::{opaque#0}`
#8 [check_well_formed] checking that `Foo::{opaque#0}` is well-formed
#9 [check_type_wf] checking that types are well-formed
#10 [analysis] running analysis passes on crate `mvce`
end of query stack
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0601, E0658.
For more information about an error, try `rustc --explain E0601`.

```

</p>
</details>

<!--
query stack:
#0 [typeck] type-checking `make_foo`
#1 [check_match] match-checking `make_foo`
#2 [mir_built] building MIR for `make_foo`
#3 [has_ffi_unwind_calls] checking if `make_foo` contains FFI-unwind calls
#4 [mir_promoted] promoting constants in MIR for `make_foo`
#5 [mir_borrowck] borrow-checking `make_foo`
#6 [type_of_opaque] computing type of opaque `Foo::{opaque#0}`
#7 [type_of] computing type of `Foo::{opaque#0}`
#8 [check_well_formed] checking that `Foo::{opaque#0}` is well-formed
#9 [check_type_wf] checking that types are well-formed
#10 [analysis] running analysis passes on crate `mvce`

-->
@rustbot label +F-type_alias_impl_trait +F-type_alias_impl_trait +WG-trait-system-refactor

old solver:
```rust
error[E0308]: mismatched types
 --> /tmp/crash.rs:5:18
  |
5 | fn make_foo() -> Foo {}
  |    --------      ^^^ expected `Vec<_>`, found `()`
  |    |
  |    implicitly returns `()` as its body has no tail or `return` expression
  |
  = note: expected struct `Vec<_>`
          found unit type `()`

error: aborting due to 1 previous error
```

--- a/compiler/rustc_trait_selection/src/error_reporting/traits/fulfillment_errors.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/traits/fulfillment_errors.rs
@@ -1607,6 +1607,7 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
                         };
 
                     if let Some(lhs) = lhs.to_alias_term()
+                        && let ty::AliasTermKind::ProjectionTy | ty::AliasTermKind::ProjectionConst = lhs.kind(self.tcx)
                         && let Some((better_type_err, expected_term)) =
                             derive_better_type_error(lhs, rhs)
                     {
@@ -1615,6 +1616,7 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
                             better_type_err,
                         )
                     } else if let Some(rhs) = rhs.to_alias_term()
+                        && let ty::AliasTermKind::ProjectionTy | ty::AliasTermKind::ProjectionConst = rhs.kind(self.tcx)
                         && let Some((better_type_err, expected_term)) =
                             derive_better_type_error(rhs, lhs)
                     {

--- a/tests/ui/type-alias-impl-trait/opaque-alias-relate-issue-151331.rs
+++ b/tests/ui/type-alias-impl-trait/opaque-alias-relate-issue-151331.rs
@@ -1,0 +1,10 @@
+//@ compile-flags: -Znext-solver=globally
+#![feature(type_alias_impl_trait)]
+
+type Foo = Vec<impl Send>;
+
+#[define_opaque(Foo)]
+fn make_foo() -> Foo {}
+//~^ ERROR type mismatch resolving
+
+fn main() {}

--- a/tests/ui/type-alias-impl-trait/opaque-alias-relate-issue-151331.stderr
+++ b/tests/ui/type-alias-impl-trait/opaque-alias-relate-issue-151331.stderr
@@ -1,0 +1,9 @@
+error[E0271]: type mismatch resolving `Foo == ()`
+  --> $DIR/opaque-alias-relate-issue-151331.rs:7:18
+   |
+LL | fn make_foo() -> Foo {}
+   |                  ^^^ types differ
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0271`.

